improve junit vintage docs about parallel execution

etrandafir93 opened the pull request:
Issue: #4321

## Overview

Improving user guide section about parallel test execution at class/methods level with JUnit Vintage.

---

I hereby agree to the terms of the [JUnit Contributor License Agreement](https://github.com/junit-team/junit5/blob/002a0052926ddee57cf90580fa49bc37e5a72427/CONTRIBUTING.md#junit-contributor-license-agreement).

---

### Definition of Done

- [ ] (n/a) There are no TODOs left in the code
- [ ] (n/a) Method [preconditions](https://junit.org/junit5/docs/snapshot/api/org.junit.platform.commons/org/junit/platform/commons/util/Preconditions.html) are checked and documented in the method's Javadoc
- [ ] (n/a) [Coding conventions](https://github.com/junit-team/junit5/blob/HEAD/CONTRIBUTING.md#coding-conventions) (e.g. for logging) have been followed
- [ ] (n/a) Change is covered by [automated tests](https://github.com/junit-team/junit5/blob/HEAD/CONTRIBUTING.md#tests) including corner cases, errors, and exception handling
- [ ] (n/a) Public API has [Javadoc](https://github.com/junit-team/junit5/blob/HEAD/CONTRIBUTING.md#javadoc) and [`@API` annotations](https://apiguardian-team.github.io/apiguardian/docs/current/api/org/apiguardian/api/API.html)
- [ ] (n/a) Change is documented in the [User Guide](https://junit.org/junit5/docs/snapshot/user-guide/) and [Release Notes](https://junit.org/junit5/docs/snapshot/user-guide/#release-notes)

YongGoose commented on documentation/src/docs/asciidoc/user-guide/migration-from-junit4.adoc:
```diff
@@ -55,21 +55,76 @@ concurrent test execution. It can be enabled and configured using the following
+----
+junit.vintage.execution.parallel.enabled=true
 junit.vintage.execution.parallel.methods=true
-junit.vintage.execution.parallel.pool-size=4
```
Why did you remove this?

YongGoose commented on documentation/src/docs/asciidoc/user-guide/migration-from-junit4.adoc:
```diff
@@ -45,8 +45,8 @@ concurrent test execution. It can be enabled and configured using the following
-  Enable/disable parallel execution (defaults to `false`). Requires opt-in for `classes`
-  or `methods` to be executed in parallel using the configuration parameters below.
+  Enable/disable parallel execution (defaults to `false`).
+Requires opt-in for `classes` or `methods` to be executed in parallel using the configuration parameters below.
```
```suggestion
  Requires opt-in for `classes` or `methods` to be executed in parallel using the configuration parameters below.
```

YongGoose commented on documentation/src/docs/asciidoc/user-guide/migration-from-junit4.adoc:
```diff
@@ -55,21 +55,76 @@ concurrent test execution. It can be enabled and configured using the following
-  Specifies the size of the thread pool to be used for parallel execution. By default, the
-  number of available processors is used.
+  Specifies the size of the thread pool to be used for parallel execution.
+By default, the  number of available processors is used.
```
```suggestion
  By default, the  number of available processors is used.
```

etrandafir93 replied:
I added the spaces, as suggested - but do we need them? 

I have added the brake lines to comply with this [sembr](https://sembr.org/) convention, but I won't have an effect on the way the text is displayed. That's why I haven't added any space at the start of the row.

YongGoose replied:
I hope this is helpful to you.

![image](https://github.com/user-attachments/assets/be13858f-2ebf-40e9-bf25-593ec12333fd)

![image](https://github.com/user-attachments/assets/8fdc7ba5-7850-4f8d-a0ab-9eec56717d05)

- https://github.com/junit-team/junit5/pull/4242#discussion_r1948051443

YongGoose commented on documentation/src/docs/asciidoc/user-guide/migration-from-junit4.adoc:
```diff
@@ -55,21 +55,76 @@ concurrent test execution. It can be enabled and configured using the following
 
 [source,properties]
 ----
 junit.vintage.execution.parallel.enabled=true
 junit.vintage.execution.parallel.classes=true
```
```suggestion
junit.vintage.execution.parallel.enabled=true
junit.vintage.execution.parallel.classes=true
junit.vintage.execution.parallel.pool-size=2
```

etrandafir93 replied:
I left this property out to allow the reader to focus on the ones that directly impact the example. 

For the given example with 2 test classes and parallelization enabled per class, we'll always use 2 threads - even if the thread pool size is larger, correct?

YongGoose replied:
> For the given example with 2 test classes and parallelization enabled per class, we'll always use 2 threads - even if the thread pool size is larger, correct?

That’s correct.

While you could explain it by saying that running FooTest and BarTest uses two threads, I believe it is more appropriate to relate it to the size of the thread pool rather than directly linking the number of tests to the number of threads. For example, when running `three tests` in parallel with a thread pool size of `two`, only `two threads` will be used, regardless of the number of tests.

What do you think? @marcphilipp

marcphilipp replied:
I think explaining one concept at a time (`classes`, `methods`, both, pool size, and then sequential execution) would be easier to follow. Therefore, I'd omit it here and introduce a new "Configuring the Pool Size" subsection before the "Sequential Execution" one.

YongGoose commented on documentation/src/docs/asciidoc/user-guide/migration-from-junit4.adoc:
```diff
@@ -55,21 +55,76 @@ concurrent test execution. It can be enabled and configured using the following
+
+[source,properties]
+----
+junit.vintage.execution.parallel.enabled=true
+junit.vintage.execution.parallel.classes=true
+junit.vintage.execution.parallel.methods=true
```
The default number of threads is `Runtime.getRuntime().availableProcessors()` unless specified otherwise.
So, it would be a good idea to include a configuration setting to define the number of threads.

- https://github.com/junit-team/junit5/blob/main/junit-vintage-engine/src/main/java/org/junit/vintage/engine/execution/VintageExecutor.java#L48

YongGoose reviewed (changes requested):
It would be great to add documentation for the following cases as well.

The document looks much better now.
Thank you! 🙂

```
junit.vintage.execution.parallel.enabled=false
junit.vintage.execution.parallel.classes=true
junit.vintage.execution.parallel.methods=true
```

etrandafir93 commented:
> junit.vintage.execution.parallel.enabled=false
> junit.vintage.execution.parallel.classes=true
> junit.vintage.execution.parallel.methods=true

@YongGoose - thanks for the feedback. I have updated the PR now addressing the comments. 
I have broken down the content into 4 short subsections, as the main section grew. 

I have added the example above as well as the case where parallel execution is enabled but no parallel execution level is defined - to highlight this misconfiguration as well. 

Let me know what you think:)

YongGoose commented on documentation/src/docs/asciidoc/user-guide/migration-from-junit4.adoc:
```diff
@@ -45,8 +45,8 @@ concurrent test execution. It can be enabled and configured using the following
-  Enable/disable parallel execution (defaults to `false`). Requires opt-in for `classes`
-  or `methods` to be executed in parallel using the configuration parameters below.
+  Enable/disable parallel execution (defaults to `false`).
+ Requires opt-in for `classes` or `methods` to be executed in parallel using the configuration parameters below.
```
There should definitely be two spaces at the start :)

> I’ve also had more feedback on documentation than on code when contributing. 😂

etrandafir93 replied:
oh 😅 fixed now. I wonder if spotless can be configured to catch this

YongGoose replied:
I’m not very familiar with `Spotless`, so I’m not sure if it provides that feature.
If it does, it might be a good idea to discuss it with Marc and consider contributing.
It looks like the file below is used for `Spotless configuration`.

- https://github.com/junit-team/junit5/blob/main/gradle/plugins/common/src/main/kotlin/junitbuild.spotless-conventions.gradle.kts

jbduncan replied:
As a former contributor to Spotless, there's nothing technical stopping us from adopting such a check. It would "just" require implementing a regex (which the Gradle plugin has native support for) or finding a JVM library or CLI tool that does the job.

However, as English uses full stops for ends of sentences, word shortenings (for example, "regarding" to "re.") and inside other words sometimes (for example, "01.01.2025"), this may be tricky to do without heuristics unless we're happy to catch all full stops followed one space or 3+ spaces.

YongGoose commented:
> @YongGoose- 피드백 감사합니다. 저는 지금 PR을 업데이트하여 댓글을 해결했습니다. 주요 섹션이 커짐에 따라 콘텐츠를 4개의 짧은 하위 섹션으로 나누었습니다.
> 
> 위의 예와 병렬 실행이 활성화되어 있지만 병렬 실행 수준이 정의되지 않은 경우를 추가하여 이러한 잘못된 구성을 강조했습니다.
> 
> 여러분의 생각을 알려주세요:)

LGTM 👍

marcphilipp commented on documentation/src/docs/asciidoc/user-guide/migration-from-junit4.adoc:
```diff
@@ -45,8 +45,8 @@ concurrent test execution. It can be enabled and configured using the following
-  Enable/disable parallel execution (defaults to `false`). Requires opt-in for `classes`
-  or `methods` to be executed in parallel using the configuration parameters below.
+  Enable/disable parallel execution (defaults to `false`).
+  Requires opt-in for `classes` or `methods` to be executed in parallel using the configuration parameters below.
```
We currently use a different convention, namely hard-wrapping at 90 characters. Please revert this change and reformat your additions accordingly.

etrandafir93 replied:
fixed now

marcphilipp commented on documentation/src/docs/asciidoc/user-guide/migration-from-junit4.adoc:
```diff
@@ -55,21 +55,122 @@ concurrent test execution. It can be enabled and configured using the following
+[[migrating-from-junit4-parallel-execution-class-and-method-level]]
+==== Full Parallelization
+
+Finally, we can also enable the parallelization at both `class` and `method` level:
```
```suggestion
Finally, we can also enable the parallelization at both class and method level:
```

etrandafir93 replied:
fixed

marcphilipp commented on documentation/src/docs/asciidoc/user-guide/migration-from-junit4.adoc:
```diff
@@ -55,21 +55,122 @@ concurrent test execution. It can be enabled and configured using the following
+junit.vintage.execution.parallel.methods=true
+----
+
+Similarly, the parallelization won't work if we only enable the parallel execution,
+but we don't specifically enable a parallel execution level:
+
+[source,properties]
+----
+junit.vintage.execution.parallel.enabled=true
+junit.vintage.execution.parallel.classes=false
+junit.vintage.execution.parallel.methods=false
+----
+
+In such cases, all the tests will be executed sequentially:
+[source,plaintext]
+----
+main - BarTest::test1
+main - BarTest::test2
+main - BarTest::test3
+
+main - FooTest::test1
+main - FooTest::test2
+main - FooTest::test3
+----
```
This is actually be considered a configuration error and produces a warning. Therefore, we should point that out and omit the example below.

```suggestion
Similarly, tests will be executed sequentially if you enable parallel execution in general
but enable neither class-level nor method-level parallelization.
```

etrandafir93 replied:
removed

marcphilipp commented on documentation/src/docs/asciidoc/user-guide/migration-from-junit4.adoc:
```diff
@@ -58,18 +58,100 @@ concurrent test execution. It can be enabled and configured using the following
+==== Parallelization at Class Level
+
+Let's assume we have two test classes `FooTest` and `BarTest` with each class containing
+three unit tests. Now, let's enable the parallel execution of test classes:
```
```suggestion
three unit tests. Now, let's enable parallel execution of test classes:
```

etrandafir93 replied:
fixed now

marcphilipp commented on documentation/src/docs/asciidoc/user-guide/migration-from-junit4.adoc:
```diff
@@ -58,18 +58,100 @@ concurrent test execution. It can be enabled and configured using the following
+----
+junit.vintage.execution.parallel.enabled=true
 junit.vintage.execution.parallel.methods=true
 junit.vintage.execution.parallel.pool-size=4
```
As per my suggestion above.

```suggestion
```

etrandafir93 replied:
fixed

marcphilipp commented on documentation/src/docs/asciidoc/user-guide/migration-from-junit4.adoc:
```diff
@@ -58,18 +58,100 @@ concurrent test execution. It can be enabled and configured using the following
+[[migrating-from-junit4-parallel-execution-method-level]]
+==== Parallelization at Method Level
+
+Alternatively, we can enable the parallel test execution at a `method` level,
```
```suggestion
Alternatively, we can enable parallel test execution at a `method` level,
```

marcphilipp commented on documentation/src/docs/asciidoc/user-guide/migration-from-junit4.adoc:
```diff
@@ -58,18 +58,100 @@ concurrent test execution. It can be enabled and configured using the following
 
-With these properties set, the `VintageTestEngine` will execute tests in parallel,
-potentially significantly reducing the overall test suite execution time.
+Therefore, the test methods from within a class will be executed in parallel, while
```
```suggestion
Therefore, the test methods within each class will be executed in parallel, while
```

marcphilipp commented on documentation/src/docs/asciidoc/user-guide/migration-from-junit4.adoc:
```diff
@@ -58,18 +58,100 @@ concurrent test execution. It can be enabled and configured using the following
+[[migrating-from-junit4-parallel-execution-class-and-method-level]]
+==== Full Parallelization
+
+Finally, we can also enable the parallelization at both class and method level:
```
```suggestion
Finally, we can also enable parallelization at both class and method level:
```

marcphilipp commented on documentation/src/docs/asciidoc/user-guide/migration-from-junit4.adoc:
```diff
@@ -58,18 +58,100 @@ concurrent test execution. It can be enabled and configured using the following
+junit.vintage.execution.parallel.pool-size=8
+----
+
+With these properties set, the `VintageTestEngine` will execute all the tests classes and
```
```suggestion
With these properties set, the `VintageTestEngine` will execute all tests classes and
```

marcphilipp commented on documentation/src/docs/asciidoc/user-guide/migration-from-junit4.adoc:
```diff
@@ -58,18 +58,100 @@ concurrent test execution. It can be enabled and configured using the following
+[[migrating-from-junit4-parallel-execution-disabled]]
+==== Sequential Execution
+
+On the other hand, if we disable the parallel execution, the `VintageTestEngine`
```
```suggestion
On the other hand, if we disable parallel execution, the `VintageTestEngine`
```

marcphilipp commented on documentation/src/docs/asciidoc/user-guide/migration-from-junit4.adoc:
```diff
@@ -58,18 +58,100 @@ concurrent test execution. It can be enabled and configured using the following
+==== Sequential Execution
+
+On the other hand, if we disable the parallel execution, the `VintageTestEngine`
+will execute the tests sequentially, regardless of the other properties:
```
```suggestion
will execute all tests sequentially, regardless of the other properties:
```

YongGoose commented:
> I think explaining one concept at a time (classes, methods, both, pool size, and then sequential execution) would be easier to follow. Therefore, I'd omit it here and introduce a new "Configuring the Pool Size" subsection before the "Sequential Execution" one.

Actually, explaining the pool size together was my own idea, as mentioned in the comment below. 😅
- https://github.com/junit-team/junit5/pull/4336#discussion_r1967079410

@etrandafir93 I apologize for any confusion it may have caused.
It seems better to explain it in a separate section, following Marc’s suggestion.

etrandafir93 commented:
> > I think explaining one concept at a time (classes, methods, both, pool size, and then sequential execution) would be easier to follow. Therefore, I'd omit it here and introduce a new "Configuring the Pool Size" subsection before the "Sequential Execution" one.
> 
> Actually, explaining the pool size together was my own idea, as mentioned in the comment below. 😅
> 
> * [Improve JUnit Vintage docs about parallel execution #4336 (comment)](https://github.com/junit-team/junit5/pull/4336#discussion_r1967079410)
> 
> @etrandafir93 I apologize for any confusion it may have caused. It seems better to explain it in a separate section, following Marc’s suggestion.

@YongGoose No worries, it was a good discussion 👍 Also, I've noticed that the pool-size parameter is used internally to define the executor's parallelism level, which is slightly different. I've added a short sentence about this.

@marcphilipp thanks for the feedback, I have applied your suggestions now, let me know what you think.

marcphilipp commented on documentation/src/docs/asciidoc/user-guide/migration-from-junit4.adoc:
```diff
@@ -58,7 +58,87 @@ concurrent test execution. It can be enabled and configured using the following
+[[migrating-from-junit4-parallel-execution-method-level]]
+==== Parallelization at Method Level
+
+Alternatively, we can enable parallel test execution at a `method` level,
+rather than the `class` level:
```
```suggestion
Alternatively, we can enable parallel test execution at a method level,
rather than the class level:
```

marcphilipp commented:
@etrandafir93 Thank you! :+1:
